File: iOS/Swordle/GameViewModel.swift
//
//  GameViewModel.swift
//  Swordle
//
//  Created by Prasanna Swaminathan on 2/14/22.
//

import Foundation

class GameViewModel {
    let word: Array<Character>
    init(withWord word: Array<Character>) {
        self.word = word
        self.guesses = []
        self.results = []
    }
    
    var guesses: [Array<Character>]
    var results: [Array<GuessResult>]
    
    func submitGuess(guess: Array<Character>) throws {
        if (guess.count != word.count) {
            throw GameError.invalidLengthError(size: guess.count)
        }
        if guess.elementsEqual(word) {
            throw GameError.finished
        }
        guesses.append(guess)
        let guessResults = zip(guess, word).map { (guessChar, wordChar) -> GuessResult in
            let result: GuessResult
            switch (guessChar) {
            case wordChar:
                result = GuessResult.correct
            case let x where word.contains(x):
                result = GuessResult.inWord
            default:
                result = GuessResult.incorrect
            }
            return result
        }
        results.append(guessResults)
    }
    
    enum GuessResult {
        case incorrect
        case inWord
        case correct
    }
}

File: Android/app/src/main/java/me/swaminathan/swordle/GameViewModel.kt
package me.swaminathan.swordle

import androidx.compose.ui.graphics.Color
import androidx.lifecycle.ViewModel

class GameViewModel(val word: CharArray, val size: Int = 5) : ViewModel() {
    private val charsInWord = word.toSet()

    init {
        if (word.size != size) throw RuntimeException("Invalid word")
    }

    internal val guesses = mutableListOf<CharArray>()
    internal val results = mutableListOf<Array<GuessResult>>()

    fun submitGuess(guess: CharArray) {
        if (guess.size != size) throw Exception("invalid guess: $guess for word size: $size")
        if (guess.contentEquals(word)) {
            TODO("need to implement win condition")
        }
        val colors = guess.zip(word).map { (guessed, actual) ->
            when {
                guessed == actual -> GuessResult.Correct
                charsInWord.contains(guessed) -> GuessResult.InWord
                else -> GuessResult.Incorrect
            }
        }
        guesses.add(guess)
        results.add(colors.toTypedArray())
    }

    enum class GuessResult {
        Incorrect, InWord, Correct;

        fun color(): Color = when (this) {
            Incorrect -> Color.Gray
            InWord -> Color.Yellow
            Correct -> Color.Green
        }
    }
}